### PR TITLE
add patch for IB registration fixes

### PR DESCRIPTION
**Pull request by quasiben**
cc @pentschev @jakirkham 

closes https://github.com/rapidsai/ucx-py/issues/425

### CODE REVIEW

**Comment by jakirkham** [comment]
Looks like the patch file wasn't included

**Comment by pentschev** [comment]
Yeah, the patch file is missing, I tried pushing it to @quasiben 's branch but I don't have permission for that. The patch is available in https://gist.github.com/pentschev/1a42434b4114fd40f664ef1d7a792bc4 .

**Comment by quasiben** [comment]
Added the missing patch file

**Comment by quasiben** [comment]
rerun tests

**Comment by quasiben** [comment]
@pentschev I think this is failing with errors we saw before on the1.7.x branch and _not_ master:

```
memory/rcache.c: In function 'ucs_rcache_get':
memory/rcache.c:394:5: error: 'end' may be used uninitialized in this function [-Werror=maybe-uninitialized]
     ucs_rcache_find_regions(rcache, *start, *end - 1, &region_list);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
memory/rcache.c:470:27: note: 'end' was declared here
     ucs_pgt_addr_t start, end;
                           ^~~
memory/rcache.c:197:5: error: 'start' may be used uninitialized in this function [-Werror=maybe-uninitialized]
     ucs_pgtable_search_range(&rcache->pgtable, from, to,
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                              ucs_rcache_region_collect_callback, list);
                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
memory/rcache.c:470:20: note: 'start' was declared here
     ucs_pgt_addr_t start, end;
                    ^~~~~
```

https://gpuci.gpuopenanalytics.com/job/rapidsai/job/conda/job/ucx/job/ucx-py-prb-release/CONDA_USERNAME=rapidsai,CUDA_VERSION=10.1,FROM_IMAGE=condaforge%2Flinux-anvil-cuda,UCX_COMMIT=v1.7.x,UCX_PROC_VERSION=1.0.0,UCX_PY_COMMIT=master,UCX_VERSION=1.7.0/4/console

**Comment by pentschev** [comment]
@quasiben I've fixed that build error, could you update the patch file with https://gist.github.com/pentschev/1a42434b4114fd40f664ef1d7a792bc4?

**Comment by quasiben** [comment]
Another build error:

```
opt/conda/conda-bld/ucx-split_1583250904297/_build_env/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: warning: libcuda.so.1, needed by /opt/conda/conda-bld/ucx-split_1583250904297/work/ucx/src/ucs/.libs/libucs.so, not found (try using -rpath or -rpath-link)
/opt/conda/conda-bld/ucx-split_1583250904297/_build_env/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: /opt/conda/conda-bld/ucx-split_1583250904297/work/ucx/src/ucs/.libs/libucs.so: undefined reference to `cuMemGetAddressRange_v2'
/opt/conda/conda-bld/ucx-split_1583250904297/_build_env/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: /opt/conda/conda-bld/ucx-split_1583250904297/work/ucx/src/ucs/.libs/libucs.so: undefined reference to `cuPointerGetAttributes'
collect2: error: ld returned 1 exit status
```

I am going to try and build locally to reproduce

**Comment by pentschev** [comment]
I believe the linking errors above are fixed with the changes to `Makefile.am`. Is there a chance that the build directory wasn't properly cleaned or the patch didn't fully apply?

**Comment by quasiben** [comment]
It's possible -- i just changed the configure command but I'll also look at cleaning the directory out

**Comment by quasiben** [comment]
still failing with what i think should be a clean install.  Note I changed the build command to use `contrib/confiugre-release` and still we are getting a linking error:

The full build command looks like the following:

```
../contrib/configure-release --build=x86_64-conda_cos6-linux-gnu --host=x86_64-conda_cos6-linux-gnu --prefix=/opt/conda/conda-bld/ucx-split_1583255228520/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla --with-sysroot --enable-cma --enable-mt --enable-numa --with-gnu-ld --with-cm --with-rdmacm --with-verbs --with-cuda=/usr/local/cuda
```

**Comment by pentschev** [comment]
Let me check that.

**Comment by pentschev** [comment]
Not sure if related, but this part doesn't look right:
```
--prefix=/opt/conda/conda-bld/ucx-split_1583255228520/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla
```

**Comment by quasiben** [comment]
I should've mentioned that.  That part is just part of the conda build process

File: recipe/ib_registration_cache.patch [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+@@ -9,9 +9,9 @@ AUTOMAKE_OPTIONS    = nostdinc # avoid collision with built-in debug.h
+ lib_LTLIBRARIES     = libucs.la
+ bin_PROGRAMS        =
+ 
+-libucs_la_CPPFLAGS = $(BASE_CPPFLAGS) -DUCX_MODULE_DIR=\"$(moduledir)\"
+-libucs_la_CFLAGS   = $(BASE_CFLAGS)
+-libucs_la_LDFLAGS  = -ldl $(NUMA_LIBS) -version-info $(SOVERSION)
++libucs_la_CPPFLAGS = $(BASE_CPPFLAGS) $(CUDA_CPPFLAGS) -DUCX_MODULE_DIR=\"$(moduledir)\"
++libucs_la_CFLAGS   = $(BASE_CFLAGS) $(CUDA_CFLAGS)
++libucs_la_LDFLAGS  = -ldl $(NUMA_LIBS) $(patsubst %, -Xlinker %, $(CUDA_LDFLAGS)) -version-info $(SOVERSION)
```
Comment (jakirkham) [inline recipe/ib_registration_cache.patch]: Do we need to change these? If not, could we ever them? If so, could you please explain why these are needed?

Comment (quasiben) [reply recipe/ib_registration_cache.patch]: These are changed recommended by Akshay but I don't explicitly know why

File: recipe/install_ucx.sh [hunk 2]
```diff
@@ -21,7 +27,9 @@ cd "${SRC_DIR}/ucx"
     --with-cm \
     --with-rdmacm \
     --with-verbs \
-    ${CUDA_CONFIG_ARG}
+    ${CUDA_CONFIG_ARG} \
+    CPPFLAGS="-I/$CUDA_HOME/include" \
```
Comment (jakirkham) [inline recipe/install_ucx.sh]: Could we drop this line as well?

File: recipe/install_ucx.sh [hunk 3]
```diff
@@ -8,8 +8,14 @@ if [ ${cuda_compiler_version} != "None" ]; then
...
 cd "${SRC_DIR}/ucx"
+# needed to clean out directory -- some patches will change
+# Makefile.am
+git clean -fdx
+cat src/ucs/Makefile.am
 ./autogen.sh
-./configure \
+mkdir -p build_ucx
+cd build_ucx
+../contrib/configure-release \
```
Comment (jakirkham) [inline recipe/install_ucx.sh]: Is there a reason we are using this instead of calling `./configure`?

**Comment by quasiben** [comment]
@jakirkham this is mostly me trying to get the build passing.  I wouldn't review all the changes here until the build works

**Comment by jakirkham** [comment]
Right. Am just trying to strip away things to see if we can get to a minimal patch with a clearer idea of what is not working. Could take over the PR instead.

**Comment by quasiben** [comment]
Let me reset this first

**Comment by jakirkham** [comment]
Ok, thanks Ben 🙂

**Comment by quasiben** [comment]
Thanks John.  All yours now

**Comment by quasiben** [comment]
@jakirkham I think we are also hitting similar failures in your recent maintainer update PR: https://github.com/rapidsai/ucx-split-feedstock/pull/16

File: recipe/ib_registration_cache.patch (hunk 1 above)
Comment (pentschev) [reply recipe/ib_registration_cache.patch]: These are needed because we're introducing CUDA code in the IB codepath.

**Comment by jakirkham** [comment]
That seems to be a different failure caused by the addition of `psutil` as a dependency. Fixing with PR ( https://github.com/rapidsai/ucx-split-feedstock/pull/17 ).

**Comment by pentschev** [comment]
> That seems to be a different failure caused by the addition of psutil as a dependency. Fixing with PR ( #17 ).

Sorry about that's probably my mistake. I also opened https://github.com/rapidsai/ucx-py/pull/426 to add it as a dependency in UCX-Py.

**Comment by jakirkham** [comment]
No need to be sorry. Thanks for fixing that bug in fact. 😉

Also thanks for listing it in UCX-Py. 🙂

**Comment by quasiben** [comment]
@jakirkham any ideas on why this still fails.  I am trying to reproduce locally but I get stuck at CUDA_HOME issues in the docker container

**Comment by jakirkham** [comment]
Not sure. Out-of-date Docker image maybe? Did you try pulling it again?

**Comment by quasiben** [comment]
i hacked the docker run command, add `/usr/local/cuda` as a volume and I am able to reproduce locally now

**Comment by jakirkham** [comment]
How are you running it? Normally I just run `python build-locally.py` and select an image.

**Comment by quasiben** [comment]
that is how i run it but it does not bring in the cuda directory for me.

**Comment by quasiben** [comment]
I wonder if UCX cares about symlinks.  The error is complain about missing a file:

> /home/conda/feedstock_root/build_artifacts/ucx-split_1583290252052/_build_env/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: warning: libcuda.so.1, needed by /home/conda/feedstock_root/bui
ld_artifacts/ucx-split_1583290252052/work/ucx/src/ucs/.libs/libucs.so, not found (try using -rpath or -rpath-link)

However, that file exists but it's a symlink

> /home/conda/feedstock_root/build_artifacts/ucx-split_1583290252052/work/ucx/src/ucs/.libs/libucs.so -> libucs.so.0.0.0

**Comment by jakirkham** [comment]
Are you running this on a machine that NVIDIA Docker?

**Comment by quasiben** [comment]
ah, that could be the problem. I am on dgx13

**Comment by jakirkham** [comment]
Yeah NVIDIA Docker is great for some things for sure. Unfortunately it does cause build issues in my experience.

The warning you saw above may be related to that as well.

Normally I just build on my laptop to solve these problems 🙂

**Comment by quasiben** [comment]
I think it's the same issue as in the gpuCI build.

**Comment by quasiben** [comment]
sorry, i should've said, NVIDIA-docker is not installed on the machine

**Comment by jakirkham** [comment]
Are you sure? I'm pretty sure the DGXs all have NVIDIA Docker.

**Comment by quasiben** [comment]
```
(base) bzaitlen@dgx13:~$ nvidia-docker
nvidia-docker: command not found
```

**Comment by jakirkham** [comment]
Yeah it used to be a separate command, but they have since baked it into Docker more properly.

```
$ docker info
...
 Runtimes: nvidia runc
...
```

**Comment by jakirkham** [comment]
Also if you run `cat /etc/docker/daemon.json` on one of the DGXs, it will show something like this...

```json
{
    "default-runtime": "nvidia",
    "runtimes": {
        "nvidia": {
            "path": "nvidia-container-runtime",
            "runtimeArgs": []
        }
    }
}
```

This means NVIDIA Docker is being used as the default runtime. Unfortunately there isn't a way to turn it off when it is already on without changing this file and restarting the Docker service (maybe even the machine).

When NVIDIA Docker is not used by default, there is at least the option to use it or not via a CLI argument.

**Comment by quasiben** [comment]
On 13 `/etc/docker/daemon.json` does not exist. But I think we are getting closer to the issue. If you try building locally I would guess you get the same error.  When i checked the `libucs.so` file I saw that it is not properly linked against `libcuda.so`

```
ldd /home/conda/feedstock_root/build_artifacts/ucx-split_1583290252052/work/ucx/src/ucs/.libs/libucs.so
        linux-vdso.so.1 =>  (0x00007ffc5994f000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f32a84a9000)
        libucm.so.0 => /home/conda/feedstock_root/build_artifacts/ucx-split_1583290252052/work/ucx/src/ucm/.libs/libucm.so.0 (0x00007f32a88e8000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f32a828c000)
        librt.so.1 => /lib64/librt.so.1 (0x00007f32a8084000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f32a7e80000)
        libcuda.so.1 => not found
```

**Comment by jakirkham** [comment]
I don't have access to that machine. So really can't say how they have configured it. From my experience this is how the DGXs in the lab are configured. Would check `docker info` to see what it says.

Yeah that is a bit of a red herring. `libcuda.so.1` is a library that ships the CUDA bits needed to work with your machines driver. This library is intentionally not included in the Docker image as it shouldn't be linked against. Instead it should be linking to the stub library at build time (using `libcuda.so.1` at runtime). Please see [this comment for details]( https://github.com/NVIDIA/nvidia-docker/issues/775#issuecomment-400035216 ).

**Comment by jakirkham** [comment]
My guess is we are linking against the wrong library.

```
.../libucs.so: undefined reference to `cuMemGetAddressRange_v2'
.../ld: .../libucs.so: undefined reference to `cuPointerGetAttributes'
```

(shortened for brevity)

**Comment by jakirkham** [comment]
Just for clarity, I have also reproduced the error locally.

**Comment by jakirkham** [comment]
As Ben pointed out offline, issue ( https://github.com/tensorflow/tensorflow/issues/13243 ) seem relevant. This seems very much related as it is chroot build issue. Trying a potential fix locally based on that.

**Comment by jakirkham** [comment]
I tried adding some other directories the rpath by adding `-Wl,rpath,<...>` to the `LDFLAGS`. Unfortunately this didn't work.

Also [tried this `ldconfig` trick]( https://github.com/tensorflow/tensorflow/issues/13243#issuecomment-347660309 ) without any luck.

**Comment by jakirkham** [comment]
The build time here is pretty rough for iterations (~10mins). So have created a smaller reproducer for myself. Should add am not too interested in code correctness (like error handling, clean up, etc.). Mostly am trying to use the same symbols and build process to get a similar error. This seems to be sufficiently minimal:

```c
#include <cuda_runtime_api.h>
#include <cuda.h>

int main() {
	float *ptr; size_t size = 10;
	cudaError_t err = cudaMalloc((void**)&ptr, size);

	CUmemorytype memtype = 0;
	void *attrdata[] = {(void *)&memtype};
	CUpointer_attribute attributes[1] = {CU_POINTER_ATTRIBUTE_MEMORY_TYPE};
	CUresult res = cuPointerGetAttributes(1, &attributes[0], attrdata, (CUdeviceptr)ptr);

	return 0;
}
```

**Pull request merged**